### ISSUE TITLE
Nextflow version 23.10.0 is incompatible with QIIME2 (pipeline fails), solution: prepend 'NXF_VER=23.04.4' when running the pipeline

### ISSUE THREAD

**Issue by bioinfogaby** [nf-core/ampliseq#654]
### Description of the bug

FYI, **ASV_seqs.fasta** is present in the work directory (b12ddbfb05b35bef6eb415cb9f1ef0). I've no clue on the origin of the error.

```
ERROR ~ Error executing process > 'NFCORE_AMPLISEQ:AMPLISEQ:QIIME2_TAXONOMY:QIIME2_INSEQ (ASV_seqs.fasta)'

Caused by:
  Process `NFCORE_AMPLISEQ:AMPLISEQ:QIIME2_TAXONOMY:QIIME2_INSEQ (ASV_seqs.fasta)` terminated with an error exit status (151)

Command executed:

  qiime tools import \
      --input-path "ASV_seqs.fasta" \
      --type 'FeatureData[Sequence]' \
      --output-path rep-seqs.qza
  
  cat <<-END_VERSIONS > versions.yml
  "NFCORE_AMPLISEQ:AMPLISEQ:QIIME2_TAXONOMY:QIIME2_INSEQ":
      qiime2: $( qiime --version | sed '1!d;s/.* //' )
  END_VERSIONS

Command exit status:
  151

Command output:
  (empty)

Command error:
  Matplotlib created a temporary config/cache directory at /tmp/matplotlib-38xyr0wz because the default path (/home/qiime2/matplotlib) is not a writable directory; it is highly recommended to set the MPLCONFIGDIR environment variable to a writable directory, in particular to speed up the import of Matplotlib and to better support multiprocessing.
  Traceback (most recent call last):
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/q2cli/builtin/tools.py", line 266, in import_data
      artifact = qiime2.sdk.Artifact.import_data(type, input_path,
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/qiime2/sdk/result.py", line 299, in import_data
      pm = qiime2.sdk.PluginManager()
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/qiime2/sdk/plugin_manager.py", line 67, in __new__
      self._init(add_plugins=add_plugins)
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/qiime2/sdk/plugin_manager.py", line 105, in _init
      plugin = entry_point.load()
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/pkg_resources/__init__.py", line 2518, in load
      return self.resolve()
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/pkg_resources/__init__.py", line 2524, in resolve
      module = __import__(self.module_name, fromlist=['__name__'], level=0)
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/q2_diversity/__init__.py", line 11, in <module>
      from ._beta import (beta, beta_phylogenetic, bioenv,
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/q2_diversity/_beta/__init__.py", line 13, in <module>
      from ._beta_rarefaction import beta_rarefaction
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/q2_diversity/_beta/_beta_rarefaction.py", line 23, in <module>
      from .._ordination import pcoa
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/q2_diversity/_ordination.py", line 20, in <module>
      import umap as up
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/umap/__init__.py", line 2, in <module>
      from .umap_ import UMAP
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/umap/umap_.py", line 41, in <module>
      from umap.layouts import (
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/umap/layouts.py", line 40, in <module>
      def rdist(x, y):
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/numba/core/decorators.py", line 234, in wrapper
      disp.enable_caching()
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/numba/core/dispatcher.py", line 863, in enable_caching
      self._cache = FunctionCache(self.py_func)
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/numba/core/caching.py", line 601, in __init__
      self._impl = self._impl_class(py_func)
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/numba/core/caching.py", line 337, in __init__
      raise RuntimeError("cannot cache function %r: no locator available "
  RuntimeError: cannot cache function 'rdist': no locator available for file '/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/umap/layouts.py'
  
  An unexpected error has occurred:
  
    cannot cache function 'rdist': no locator available for file '/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/umap/layouts.py'
  
  See above for debug info.

Work dir:
  /mnt/DATA/bioinfo/projects/AmpliseqDir/work/e5/b12ddbfb05b35bef6eb415cb9f1ef0

Tip: you can try to figure out what's wrong by changing to the process work dir and showing the script file named `.command.sh`

 -- Check '.nextflow.log' file for details
```

### Command used and terminal output

```console
$ nextflow run nf-core/ampliseq -r '2.7.0' -profile singularity --input ' /mnt/DATA/bioinfo/projects/AmpliseqDir/sampleSheet_16S.tsv' --FW_primer 'CCTACGGGNGGCWGCAG' --RV_primer 'GACTACHVGGGTATCTAATCC' --metadata ' /mnt/DATA/bioinfo/projects/AmpliseqDir/metadata.tsv' --outdir ampliseq_results_16S_2.7.0 --max_cpus 4 --max_memory 15GB
```

### System information

Nextflow version:  23.10.0 build 5889
Hardware: Desktop
Executor: local
Container engine: Singularity
OS: Linux
Version of nf-core/ampliseq: 2.7.0

**Comment by bioinfogaby** [comment]
Yeap, that worked. Thanks!

**Comment by muffato** [comment]
Nextflow 23.10 adds the `--no-home` option when using Singularity. Maybe this tool wanted to cache data under the home directory ?

**Comment by d4straub** [comment]
Hm quite possible. How to test that?
- docker instead of singularity should work
- make the tool use another cache
Correct?

**Comment by MatthiasZepper** [comment]
> Nextflow 23.10 adds the `--no-home` option when using Singularity. Maybe this tool wanted to cache data under the home directory ?

That is exactly what the error message states, no?

```
Matplotlib created a temporary config/cache directory at /tmp/matplotlib-38xyr0wz 
because the default path (/home/qiime2/matplotlib) is not a writable directory; 
it is highly recommended to set the MPLCONFIGDIR environment variable to a writable directory, 
in particular to speed up the import of Matplotlib and to better support multiprocessing.
```

**Comment by d4straub** [comment]
I just wanted to confirm the problem on my laptop (to later run with docker; I have installed singularity & docker) with
`NXF_VER=23.10.0 nextflow run nf-core/ampliseq -r 2.7.0 -profile test,singularity --outdir results_test_2-7-0 -resume`
and it _succeeded_. So it seems to differ between systems.

So I tried all 3 systems that I have available at the moment:

**Failing system (Workstation)**
* Nextflow version: 23.10.0
* Java version: openjdk 17.0.9 2023-10-17 LTS, openJDK Runtime Environment (Red_Hat-17.0.9.0.9-1) (build 17.0.9+9-LTS)
* Operating system: Linux 5.14.0-284.30.1.el9_2.x86_64 x86_64
* Bash version: GNU bash, version 5.1.8(1)-release (x86_64-redhat-linux-gnu)
* Singularity: singularity version 3.8.7-1.el9

**Succeeding system (Laptop):** 
* Nextflow version: 23.10.0
* Java version: openjdk 11.0.20.1 2023-08-24, OpenJDK Runtime Environment (build 11.0.20.1+1-post-Ubuntu-0ubuntu120.04)
* Operating system: Linux 5.15.0-87-generic x86_64
* Bash version: GNU bash, version 5.0.17(1)-release (x86_64-pc-linux-gnu)
* Singularity: singularity-ce version 3.10.5-focal

**Succeeding system (hpc):** 
* Nextflow version: 23.10.0
* Java version: openjdk 17.0.5 2022-10-18 LTS, OpenJDK Runtime Environment (Red_Hat-17.0.5.0.8-2.el8_6) (build 17.0.5+8-LTS)
* Operating system: Linux 4.18.0-372.32.1.el8_6.x86_64 x86_64
* Bash version: GNU bash, version 4.4.20(1)-release (x86_64-redhat-linux-gnu)
* Singularity: singularity version 3.8.7-1.el8

**Comment by muffato** [comment]
> That is exactly what the error message states, no?
> 
> ```
> Matplotlib created a temporary config/cache directory at /tmp/matplotlib-38xyr0wz 
> because the default path (/home/qiime2/matplotlib) is not a writable directory; 
> it is highly recommended to set the MPLCONFIGDIR environment variable to a writable directory, 
> in particular to speed up the import of Matplotlib and to better support multiprocessing.
> ```
> 
> In previous versions, `/home` was mounted inside the container and writeable, so `/home/qiime2/matplotlib` could be used for caching.

Great. In that case, I would imagine two workarounds:
- In the module, do something like `export MPLCONFIGDIR=$PWD` to keep the cache local to the work directory, which is already mounted read+write
- Add an optional input channel to the module for a cache directory, which Nextflow would properly stage and mount, and use it in `export MPLCONFIGDIR` (and default to `$PWD`)

**Comment by d4straub** [comment]
Thanks @muffato & @MatthiasZepper for your intrest and suggestions!
I added `export MPLCONFIGDIR="${PWD}/HOME"` to all processes that use QIIME2, and indeed the part with problematic MPLCONFIGDIR is solved, however the process still fails. Matplotlib wasnt the root of the problem it seems to me.
Here is the complete error message:

```
ERROR ~ Error executing process > 'NFCORE_AMPLISEQ:AMPLISEQ:QIIME2_PREPTAX:QIIME2_EXTRACT (GTGYCAGCMGCCGCGGTAA-GGACTACNVGGGTWTCTAAT)'

Caused by:
  Process `NFCORE_AMPLISEQ:AMPLISEQ:QIIME2_PREPTAX:QIIME2_EXTRACT (GTGYCAGCMGCCGCGGTAA-GGACTACNVGGGTWTCTAAT)` terminated with an error exit status (1)

Command executed:

  export XDG_CONFIG_HOME="${PWD}/HOME"
  export MPLCONFIGDIR="${PWD}/HOME"
  
     ### Import
     qiime tools import \
         --type 'FeatureData[Sequence]' \
         --input-path greengenes85.fna \
         --output-path ref-seq.qza
     qiime tools import \
         --type 'FeatureData[Taxonomy]' \
         --input-format HeaderlessTSVTaxonomyFormat \
         --input-path greengenes85.tax \
         --output-path ref-taxonomy.qza
     #Extract sequences based on primers
     qiime feature-classifier extract-reads \
         --i-sequences ref-seq.qza \
         --p-f-primer GTGYCAGCMGCCGCGGTAA \
         --p-r-primer GGACTACNVGGGTWTCTAAT \
         --o-reads GTGYCAGCMGCCGCGGTAA-GGACTACNVGGGTWTCTAAT-ref-seq.qza \
         --quiet
  
     cat <<-END_VERSIONS > versions.yml
     "NFCORE_AMPLISEQ:AMPLISEQ:QIIME2_PREPTAX:QIIME2_EXTRACT":
         qiime2: $( qiime --version | sed '1!d;s/.* //' )
     END_VERSIONS

Command exit status:
  1

Command output:
  (empty)

Command error:
  Traceback (most recent call last):
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/q2cli/builtin/tools.py", line 266, in import_data
      artifact = qiime2.sdk.Artifact.import_data(type, input_path,
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/qiime2/sdk/result.py", line 299, in import_data
      pm = qiime2.sdk.PluginManager()
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/qiime2/sdk/plugin_manager.py", line 67, in __new__
      self._init(add_plugins=add_plugins)
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/qiime2/sdk/plugin_manager.py", line 105, in _init
      plugin = entry_point.load()
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/pkg_resources/__init__.py", line 2518, in load
      return self.resolve()
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/pkg_resources/__init__.py", line 2524, in resolve
      module = __import__(self.module_name, fromlist=['__name__'], level=0)
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/q2_diversity/__init__.py", line 11, in <module>
      from ._beta import (beta, beta_phylogenetic, bioenv,
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/q2_diversity/_beta/__init__.py", line 13, in <module>
      from ._beta_rarefaction import beta_rarefaction
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/q2_diversity/_beta/_beta_rarefaction.py", line 23, in <module>
      from .._ordination import pcoa
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/q2_diversity/_ordination.py", line 20, in <module>
      import umap as up
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/umap/__init__.py", line 2, in <module>
      from .umap_ import UMAP
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/umap/umap_.py", line 41, in <module>
      from umap.layouts import (
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/umap/layouts.py", line 40, in <module>
      def rdist(x, y):
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/numba/core/decorators.py", line 234, in wrapper
      disp.enable_caching()
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/numba/core/dispatcher.py", line 863, in enable_caching
      self._cache = FunctionCache(self.py_func)
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/numba/core/caching.py", line 601, in __init__
      self._impl = self._impl_class(py_func)
    File "/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/numba/core/caching.py", line 337, in __init__
      raise RuntimeError("cannot cache function %r: no locator available "
  RuntimeError: cannot cache function 'rdist': no locator available for file '/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/umap/layouts.py'
  
  An unexpected error has occurred:
  
    cannot cache function 'rdist': no locator available for file '/opt/conda/envs/qiime2-2023.7/lib/python3.8/site-packages/umap/layouts.py'
  
  See above for debug info.

Work dir:
  /home/bcgsd01/test_ampliseq/work/08/8a446d3afbec175518f6be0768a7d6

Tip: when you have fixed the problem you can continue the execution adding the option `-resume` to the run command line

 -- Check '.nextflow.log' file for details
```

**Comment by MatthiasZepper** [comment]
I think `/tmp` is the better choice nonetheless. What happens with this setting?

```
export MPLCONFIGDIR="/tmp/mplconfigdir"
export NUMBA_CACHE_DIR="/tmp/numbacache"
```

**Comment by d4straub** [comment]
Great, thanks, that did work indeed!
Why would `/tmp` be better? Is that common practice in nf-core (I am not aware of an example, do you)? I just worry a bit about clouds (cannot test that, except AWS test).

**Comment by d4straub** [comment]
I checked for examples in pipelines and nf-core/modules and found only very few examples:

nf-core/modules:
- [export ALEVIN_FRY_HOME=.](https://github.com/nf-core/modules/blob/1e729660254e84bf0b6451a94b3ca511b6a3d1e4/modules/nf-core/simpleaf/index/main.nf#L33)
- [export TEMP='./'](https://github.com/nf-core/modules/blob/1e729660254e84bf0b6451a94b3ca511b6a3d1e4/modules/nf-core/tcoffee/seqreformat/main.nf#L25)
- [export _JAVA_OPTIONS=-Djava.io.tmpdir=./tmp](https://github.com/nf-core/modules/blob/1e729660254e84bf0b6451a94b3ca511b6a3d1e4/modules/nf-core/qualimap/rnaseq/main.nf#L36)

nf-core/marsseq:
- [export TMPDIR=/tmp](https://github.com/nf-core/marsseq/blob/b02ede773092f8a8a5999400acda6894d51b4d26/modules/local/qc/report/main.nf#L33)

nf-core/rnafusion
- [export TMPDIR=/tmp](https://github.com/nf-core/rnafusion/blob/cc52314b4d4b32b8b0d2a545151f6c86be6a9892/modules/local/starfusion/build/main.nf#L17)

So nf-core modules seem to put tmp and home into the work dir `.`, while the two pipelines use `/tmp`. This pipeline uses `export XDG_CONFIG_HOME="\${PWD}/HOME"`, e.g.  [here](https://github.com/nf-core/ampliseq/blob/4e48b7100302e2576ac1be2ccc7d464253e9d20e/modules/local/qiime2_extract.nf#L25).

Not sure what to conclude here, so different solutions :)

**Comment by muffato** [comment]
Another example for your list, @d4straub : `./tmp` is this [module](https://github.com/nf-core/modules/pull/3914). The reason for choosing `./tmp` instead of letting the tool use `/tmp` is that when the job is killed on a HPC, it would then leave some leftover files on `/tmp` which take space and may get in the way of other users / processes. By keeping the files local to `./tmp`, it's fully within the user's own work directory, which the user can easily clean up.

In your case, and I know nothing about the tool so I'm just guessing, if it's a "cache", then presumably the tool will not clean it up at the end, since the purpose of a cache is to keep files around for the next run. Using `/tmp` and running the pipeline on a HPC would mean that likely, every run would hit a different compute node and the cluster may accumulate caches on all compute nodes over time ! That is not the purpose of `/tmp` :)

So either consider it a purely temporary necessity and use `./tmp`, or make it a proper pipeline parameter that the user can set to wherever they want and reuse between runs. (by the way, would it even make sense to share the cache between pipeline runs ?)

**Comment by d4straub** [comment]
Thanks for that great explanation! I'll test `./tmp` to make sure it'll work fine.

I researched where the older export code is coming from in ampliseq and I found the addition of it in https://github.com/nf-core/ampliseq/pull/163. In that PR was a change from `export HOME=./HOME` to `export HOME="\${PWD}/HOME"` [here](https://github.com/nf-core/ampliseq/commit/26f05831f5be1cd85d5fb1c8cc6e9fdae09b6cc0), because @skrakau suggested "Better not use relative paths". Since there are some examples now with relative paths and I found none with absolute paths, I am preferring the relative paths. Or are you aware of any problems Sabrina?

**Comment by MatthiasZepper** [comment]
I think the nf-core #modules channel would be the appropriate place to get some more input on the issue.

I am certainly not an expert in this matter, but @muffato  's explanation strikes me as strange. `/tmp` [is the default path within Linux file system structure for temporary files](https://www.geeksforgeeks.org/linux-directory-structure). Its sole purpose is to offer applications a dedicated place to store temporary files. These files are generally deleted whenever the system is restarted and may be deleted at any time by utilities such as _tmpwatch_. On HPC systems, that are not rebooted frequently, `/tmp` is usually configured to mount a specific scratch file system that is better suited than the ordinary distributed file systems for quick caching and updates. 

The reason why you do not find many modules or pipelines with explicit configuration is, that many tools just use `/tmp` for caching files and the respective container technologies mount the hosts `/tmp` there. [All the respective config happens on the profile level](https://github.com/search?q=repo%3Anf-core%2Fconfigs+%2Ftmp&type=code) and Nextflow has corresponding config options for the different executors (e.g. [for AWS](https://nextflow.io/docs/latest/aws.html#volume-mounts)) and container technologies.

In summary, I advocate using `/tmp`, since it is the path that is specifically meant for this purpose and also configured accordingly on the different executors.

**Comment by d4straub** [comment]
Thanks a lot for your input!
I did get the feedback however that `/tmp` is not a good choice for example on our hpc because it is using a scratch system that is not necessarily connected to `/tmp`. Therefore I was advised to use `./tmp` or similar (which will then use the scratch system as intended). I did check the size of those folder that I want to redirect and its just a couple of MB at most. So I'm going to use `./<folder>` for now.
